File: matterbase/content.py
"""Content extraction and rendering helpers — no Textual dependency.

Ported from matterbase-next, trimmed to the unified-view source types:
markdown / typst (YAML-block records) and jsonl (line records).
"""

import os
import re
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

import yaml


# ---------------------------------------------------------------------------
# Source typing — drives the adaptive preview
# ---------------------------------------------------------------------------

MARKDOWN_SUFFIXES = {".md", ".markdown", ""}
TYPST_SUFFIXES = {".typ", ".typst"}


def source_type(path: str) -> str:
    """Classify a record's source file: 'markdown', 'typst', 'jsonl', 'other'."""
    suffix = Path(path).suffix.lower()
    if suffix in MARKDOWN_SUFFIXES:
        return "markdown"
    if suffix in TYPST_SUFFIXES:
        return "typst"
    if suffix == ".jsonl":
        return "jsonl"
    return "other"


# ---------------------------------------------------------------------------
# MMD header parsing
# ---------------------------------------------------------------------------

_MMD_KEY_RE = re.compile(r'^[\w. -]+\s*:')


def split_mmd_header(content: str) -> tuple[str, str]:
    """If content starts with MMD key: value pairs, return (yaml_lines, body).

    Returns ("", content) if no MMD header found.
    """
    lines = content.split("\n")
    mmd_lines: list[str] = []
    for line in lines:
        if not line.strip():
            break
        if _MMD_KEY_RE.match(line):
            mmd_lines.append(line)
        else:
            break
    if not mmd_lines:
        return "", content
    rest_start = len(mmd_lines)
    if rest_start < len(lines) and not lines[rest_start].strip():
        rest_start += 1
    return "\n".join(mmd_lines), "\n".join(lines[rest_start:])


# ---------------------------------------------------------------------------
# Markdown section extraction
# ---------------------------------------------------------------------------

def extract_markdown_section(content: str, target_block_start: int) -> str:
    """Return the markdown section (nearest heading → next same-level heading)
    that contains the line at *target_block_start*."""
    lines = content.splitlines()

    heading_line_idx = 0
    heading_level = 0
    for i in range(target_block_start - 1, -1, -1):
        m = re.match(r"^(#{1,6})\s", lines[i])
        if m:
            heading_line_idx = i
            heading_level = len(m.group(1))
            break

    section_end_idx = len(lines)
    if heading_level > 0:
        stop_pattern = re.compile(r"^#{1," + str(heading_level) + r"}\s")
        for i in range(heading_line_idx + 1, len(lines)):
            if stop_pattern.match(lines[i]):
                section_end_idx = i
                break

    return "\n".join(lines[heading_line_idx:section_end_idx]).rstrip()


def extract_section_for_record(
    body: str,
    record: dict,
    frontmatter_keys: set[str] | None = None,
) -> str:
    """Return the markdown section containing the YAML block matching *record*.

    frontmatter_keys: excluded when building the match fingerprint so that
    frontmatter-only fields don't prevent matching.
    """
    lines = body.splitlines()
    fm_keys = frontmatter_keys or set()

    match_fields: dict[str, str] = {}
    if record.get("id") is not None and "id" not in fm_keys:
        match_fields["id"] = str(record["id"])
    else:
        for k, v in record.items():
            if not k.startswith("_") and k not in fm_keys and v is not None:
                match_fields[k] = str(v)

    if not match_fields:
        return ""

    block_start = None
    i = 0
    while i < len(lines):
        if re.match(r"^```ya?ml\s*$", lines[i]):
            j = i + 1
            while j < len(lines) and not re.match(r"^```\s*$", lines[j]):
                j += 1
            try:
                block_data = yaml.safe_load("\n".join(lines[i + 1 : j])) or {}
                if all(str(block_data.get(k, "")) == v for k, v in match_fields.items()):
                    block_start = i
                    break
            except yaml.YAMLError:
                pass
            i = j + 1
        else:
            i += 1

    if block_start is None:
        return ""

    return extract_markdown_section(body, block_start)


def split_frontmatter(content: str, mmd: bool = False) -> tuple[str, str, set[str]]:
    """Split *content* into (frontmatter_yaml, body, frontmatter_keys).

    frontmatter_yaml is the raw YAML text without delimiters ("" when absent).
    Falls back to MMD header parsing when *mmd* is set.
    """
    if content.startswith("---"):
        sections = content.split("---", 2)
        if len(sections) >= 3:
            fm = sections[1]
            body = sections[2]
            try:
                fm_data = yaml.safe_load(fm) or {}
                keys = set(fm_data.keys()) if isinstance(fm_data, dict) else set()
            except yaml.YAMLError:
                keys = set()
            return fm.strip("\n"), body, keys
    if mmd:
        mmd_yaml, mmd_body = split_mmd_header(content)
        if mmd_yaml:
            keys = {
                line.split(":", 1)[0].strip()
                for line in mmd_yaml.splitlines()
                if ":" in line
            }
            return mmd_yaml, mmd_body, keys
    return "", content, set()


# ---------------------------------------------------------------------------
# Compact content extraction (frontmatter + YAML blocks with headings)
# ---------------------------------------------------------------------------

def extract_compact_content(
    path: str,
    compact_tasks_heading: str = "Tasks",
    mmd: bool = False,
) -> str:
    """Extract frontmatter + YAML code blocks (with preceding heading) + Tasks section."""
    try:
        content = Path(path).read_text(encoding="utf-8", errors="replace")
    except OSError:
        return ""

    parts: list[str] = []
    fm, body, _ = split_frontmatter(content, mmd=mmd)
    if fm:
        parts.append(f"---\n{fm}\n---")

    lines = body.splitlines()

    i = 0
    pending_heading: str | None = None
    while i < len(lines):
        line = lines[i]
        if re.match(r"^#{1,6}\s", line):
            pending_heading = line
            i += 1
            continue
        if re.match(r"^```ya?ml\s*$", line):
            k = i + 1
            while k < len(lines) and not re.match(r"^```\s*$", lines[k]):
                k += 1
            block: list[str] = []
            if pending_heading:
                block.append(pending_heading)
            block.extend(lines[i : k + 1])
            parts.append("\n".join(block))
            pending_heading = None
            i = k + 1
            continue
        if line.strip():
            pending_heading = None
        i += 1

    tasks_lines: list[str] = []
    in_tasks = False
    for line in lines:
        if re.match(r"^## " + re.escape(compact_tasks_heading) + r"\b", line):
            in_tasks = True
            tasks_lines = [line]
        elif in_tasks:
            if re.match(r"^## ", line):
                break
            tasks_lines.append(line)
    if tasks_lines:
        parts.append("\n".join(tasks_lines).rstrip())

    return "\n\n".join(parts)


# ---------------------------------------------------------------------------
# apex / bat rendering
# ---------------------------------------------------------------------------

@dataclass
class ApexConfig:
    theme: str = ""
    width: int = 0
    code_highlight: str = ""
    code_highlight_theme: str = ""


def render_with_apex(
    path: str,
    cfg: ApexConfig,
    tmp_content: str | None = None,
) -> str:
    """Render *path* (or *tmp_content* written to a tempfile) via apex.

    Returns ANSI-escaped string, or "" if apex is unavailable or fails.
    tmp_content: if given, write to a .md tempfile and pass that to apex instead.
    """
    tmp_path: str | None = None
    try:
        if tmp_content is not None:
            with tempfile.NamedTemporaryFile(
                mode="w", suffix=".md", delete=False, encoding="utf-8"
            ) as f:
                f.write(tmp_content)
                tmp_path = f.name
        render_path = tmp_path if tmp_path else path
        cmd = ["apex", render_path, "--plugins", "-t", "terminal256"]
        if cfg.code_highlight:
            cmd += ["--code-highlight", cfg.code_highlight]
        if cfg.code_highlight_theme:
            cmd += ["--code-highlight-theme", cfg.code_highlight_theme]
        if cfg.theme:
            cmd += ["--theme", cfg.theme]
        if cfg.width:
            cmd += ["--width", str(cfg.width)]
        env = os.environ.copy()
        env.setdefault("TERM", "xterm-256color")
        env.setdefault("COLORTERM", "truecolor")
        result = subprocess.run(cmd, capture_output=True, timeout=10, env=env)
        if result.returncode == 0:
            return result.stdout.decode("utf-8", errors="replace")
        return ""
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return ""
    finally:
        if tmp_path:
            try:
                os.unlink(tmp_path)
            except OSError:
                pass


def render_with_bat(path: str) -> str:
    """Syntax-highlight *path* via bat. Returns ANSI string or ""."""
    try:
        env = os.environ.copy()
        env.setdefault("TERM", "xterm-256color")
        env.setdefault("COLORTERM", "truecolor")
        result = subprocess.run(
            ["bat", "--color=always", "--style=plain", "--paging=never", path],
            capture_output=True,
            timeout=5,
            env=env,
        )
        if result.returncode == 0:
            return result.stdout.decode("utf-8", errors="replace")
        return ""
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return ""

File: cmd/matterbase/content.go
// Content extraction and rendering — port of matterbase's content.py.
// Pure functions plus apex/bat subprocess rendering; no TUI dependency.
package main

import (
	"context"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode"

	"gopkg.in/yaml.v3"

	bkexec "github.com/rhsev/matterbase/basekit/exec"
)

var (
	markdownSuffixes = map[string]bool{".md": true, ".markdown": true, "": true}
	typstSuffixes    = map[string]bool{".typ": true, ".typst": true}
)

// sourceType classifies a record's source file, driving the adaptive
// preview: "markdown", "typst", "jsonl", or "other".
func sourceType(path string) string {
	suffix := strings.ToLower(filepath.Ext(path))
	switch {
	case markdownSuffixes[suffix]:
		return "markdown"
	case typstSuffixes[suffix]:
		return "typst"
	case suffix == ".jsonl":
		return "jsonl"
	default:
		return "other"
	}
}

var mmdKeyRe = regexp.MustCompile(`^[\w. -]+\s*:`)

// splitMMDHeader returns (yamlLines, body) if content starts with MMD
// key: value pairs, or ("", content) if no MMD header is found.
func splitMMDHeader(content string) (string, string) {
	lines := strings.Split(content, "\n")
	var mmdLines []string
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			break
		}
		if !mmdKeyRe.MatchString(line) {
			break
		}
		mmdLines = append(mmdLines, line)
	}
	if len(mmdLines) == 0 {
		return "", content
	}
	restStart := len(mmdLines)
	if restStart < len(lines) && strings.TrimSpace(lines[restStart]) == "" {
		restStart++
	}
	return strings.Join(mmdLines, "\n"), strings.Join(lines[restStart:], "\n")
}

var headingRe = regexp.MustCompile(`^(#{1,6})\s`)

// extractMarkdownSection returns the section (nearest heading → next
// same-level heading) containing the line at targetBlockStart.
func extractMarkdownSection(content string, targetBlockStart int) string {
	lines := strings.Split(content, "\n")
	headingLineIdx, headingLevel := 0, 0
	for i := targetBlockStart - 1; i >= 0; i-- {
		if m := headingRe.FindStringSubmatch(lines[i]); m != nil {
			headingLineIdx = i
			headingLevel = len(m[1])
			break
		}
	}
	sectionEndIdx := len(lines)
	if headingLevel > 0 {
		stop := regexp.MustCompile(fmt.Sprintf(`^#{1,%d}\s`, headingLevel))
		for i := headingLineIdx + 1; i < len(lines); i++ {
			if stop.MatchString(lines[i]) {
				sectionEndIdx = i
				break
			}
		}
	}
	return strings.TrimRightFunc(strings.Join(lines[headingLineIdx:sectionEndIdx], "\n"), unicode.IsSpace)
}

// pyStr renders a YAML/JSON-decoded value the way Python's str() would,
// close enough for the fingerprint matching in extractSectionForRecord.
func pyStr(v any) string {
	switch x := v.(type) {
	case nil:
		return "None"
	case bool:
		if x {
			return "True"
		}
		return "False"
	case string:
		return x
	case int:
		return strconv.Itoa(x)
	case int64:
		return strconv.FormatInt(x, 10)
	case float64:
		if x == math.Trunc(x) && !math.IsInf(x, 0) {
			return strconv.FormatFloat(x, 'f', 1, 64)
		}
		return strconv.FormatFloat(x, 'g', -1, 64)
	default:
		return fmt.Sprintf("%v", x)
	}
}

var (
	yamlFenceOpenRe  = regexp.MustCompile("^```ya?ml\\s*$")
	yamlFenceCloseRe = regexp.MustCompile("^```\\s*$")
)

// extractSectionForRecord returns the markdown section containing the
// YAML block matching record, fingerprinted by id (or, absent an id, by
// every non-underscore, non-frontmatter field).
func extractSectionForRecord(body string, record Record, frontmatterKeys map[string]bool) string {
	lines := strings.Split(body, "\n")
	matchFields := map[string]string{}
	if id, ok := record["id"]; ok && id != nil && !frontmatterKeys["id"] {
		matchFields["id"] = pyStr(id)
	} else {
		for k, v := range record {
			if strings.HasPrefix(k, "_") || frontmatterKeys[k] || v == nil {
				continue
			}
			matchFields[k] = pyStr(v)
		}
	}
	if len(matchFields) == 0 {
		return ""
	}

	blockStart := -1
	i := 0
	for i < len(lines) {
		if !yamlFenceOpenRe.MatchString(lines[i]) {
			i++
			continue
		}
		j := i + 1
		for j < len(lines) && !yamlFenceCloseRe.MatchString(lines[j]) {
			j++
		}
		var blockData map[string]any
		if err := yaml.Unmarshal([]byte(strings.Join(lines[i+1:j], "\n")), &blockData); err == nil {
			match := true
			for k, want := range matchFields {
				got := ""
				if v, ok := blockData[k]; ok {
					got = pyStr(v)
				}
				if got != want {
					match = false
					break
				}
			}
			if match {
				blockStart = i
				break
			}
		}
		i = j + 1
	}
	if blockStart < 0 {
		return ""
	}
	return extractMarkdownSection(body, blockStart)
}

// splitFrontmatter splits content into (frontmatterYAML, body, keys).
// Falls back to MMD header parsing when mmd is set.
func splitFrontmatter(content string, mmd bool) (string, string, map[string]bool) {
	if strings.HasPrefix(content, "---") {
		sections := strings.SplitN(content, "---", 3)
		if len(sections) >= 3 {
			fm, body := sections[1], sections[2]
			keys := map[string]bool{}
			var fmData map[string]any
			if err := yaml.Unmarshal([]byte(fm), &fmData); err == nil {
				for k := range fmData {
					keys[k] = true
				}
			}
			return strings.Trim(fm, "\n"), body, keys
		}
	}
	if mmd {
		mmdYAML, mmdBody := splitMMDHeader(content)
		if mmdYAML != "" {
			keys := map[string]bool{}
			for _, line := range strings.Split(mmdYAML, "\n") {
				if idx := strings.Index(line, ":"); idx >= 0 {
					keys[strings.TrimSpace(line[:idx])] = true
				}
			}
			return mmdYAML, mmdBody, keys
		}
	}
	return "", content, map[string]bool{}
}

// extractCompactContent extracts frontmatter + YAML blocks (with their
// preceding heading) + the Tasks section — the "compact" preview.
func extractCompactContent(path, compactTasksHeading string, mmd bool) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	content := string(data)

	var parts []string
	fm, body, _ := splitFrontmatter(content, mmd)
	if fm != "" {
		parts = append(parts, "---\n"+fm+"\n---")
	}

	lines := strings.Split(body, "\n")
	i := 0
	pendingHeading := ""
	for i < len(lines) {
		line := lines[i]
		if headingRe.MatchString(line) {
			pendingHeading = line
			i++
			continue
		}
		if yamlFenceOpenRe.MatchString(line) {
			k := i + 1
			for k < len(lines) && !yamlFenceCloseRe.MatchString(lines[k]) {
				k++
			}
			var block []string
			if pendingHeading != "" {
				block = append(block, pendingHeading)
			}
			block = append(block, lines[i:min(k+1, len(lines))]...)
			parts = append(parts, strings.Join(block, "\n"))
			pendingHeading = ""
			i = k + 1
			continue
		}
		if strings.TrimSpace(line) != "" {
			pendingHeading = ""
		}
		i++
	}

	tasksHeadingRe := regexp.MustCompile(`^## ` + regexp.QuoteMeta(compactTasksHeading) + `\b`)
	h2Re := regexp.MustCompile(`^## `)
	var tasksLines []string
	inTasks := false
	for _, line := range lines {
		switch {
		case tasksHeadingRe.MatchString(line):
			inTasks = true
			tasksLines = []string{line}
		case inTasks && h2Re.MatchString(line):
			inTasks = false
		case inTasks:
			tasksLines = append(tasksLines, line)
		}
	}
	if len(tasksLines) > 0 {
		parts = append(parts, strings.TrimRightFunc(strings.Join(tasksLines, "\n"), unicode.IsSpace))
	}
	return strings.Join(parts, "\n\n")
}

// ApexConfig mirrors matterbase's apex_* config keys.
type ApexConfig struct {
	Theme              string
	Width              int
	CodeHighlight      string
	CodeHighlightTheme string
}

func withTermEnv() []string {
	env := os.Environ()
	has := func(key string) bool {
		prefix := key + "="
		for _, e := range env {
			if strings.HasPrefix(e, prefix) {
				return true
			}
		}
		return false
	}
	if !has("TERM") {
		env = append(env, "TERM=xterm-256color")
	}
	if !has("COLORTERM") {
		env = append(env, "COLORTERM=truecolor")
	}
	return env
}

// renderWithApex renders path (or tmpContent, written to a .md tempfile
// when non-nil) via apex. Returns ANSI text, or "" if apex is
// unavailable or fails.
func renderWithApex(path string, cfg ApexConfig, tmpContent *string) string {
	renderPath := path
	if tmpContent != nil {
		f, err := os.CreateTemp("", "matterbase-*.md")
		if err != nil {
			return ""
		}
		defer os.Remove(f.Name())
		defer f.Close()
		if _, err := f.WriteString(*tmpContent); err != nil {
			return ""
		}
		renderPath = f.Name()
	}
	args := []string{renderPath, "--plugins", "-t", "terminal256"}
	if cfg.CodeHighlight != "" {
		args = append(args, "--code-highlight", cfg.CodeHighlight)
	}
	if cfg.CodeHighlightTheme != "" {
		args = append(args, "--code-highlight-theme", cfg.CodeHighlightTheme)
	}
	if cfg.Theme != "" {
		args = append(args, "--theme", cfg.Theme)
	}
	if cfg.Width > 0 {
		args = append(args, "--width", strconv.Itoa(cfg.Width))
	}
	out, err := bkexec.Run(context.Background(), 10*time.Second, withTermEnv(), "apex", args...)
	if err != nil {
		return ""
	}
	return string(out)
}

// renderWithBat syntax-highlights path via bat. Returns ANSI text or "".
func renderWithBat(path string) string {
	out, err := bkexec.Run(context.Background(), 5*time.Second, withTermEnv(),
		"bat", "--color=always", "--style=plain", "--paging=never", path)
	if err != nil {
		return ""
	}
	return string(out)
}
